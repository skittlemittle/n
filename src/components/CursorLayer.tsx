import React from "react";
import Cursor from "./styles/Cursor";

interface CursorState {
  shown: boolean;
}

interface CursorProps {
  position: { row: number; column: number };
  size: { height: number; width: number };
  block: boolean;
  /** id of parent buffer */
  parent: string;
  forwardedRef: React.RefObject<HTMLDivElement>;
}

/** Renders a text cursor */
class CursorLayer extends React.Component<CursorProps, CursorState> {
  private blinky: any; // not a word about this
  constructor(props: CursorProps) {
    super(props);
    this.state = { shown: true };
  }

  componentDidMount() {
    this.blinky = setInterval(() => {
      const s = this.state.shown;
      this.setState({ shown: !s });
    }, 400);
  }

  componentWillUnmount() {
    clearInterval(this.blinky);
  }

  render() {
    const offsets = document
      .getElementById(this.props.parent)
      ?.getBoundingClientRect();
    let top = 0;
    let left = 0;
    if (offsets) {
      top = offsets.top;
      left = offsets.left;
    }
    return (
      <Cursor
        dissapear={this.state.shown}
        block={this.props.block}
        h={this.props.size.height}
        w={this.props.size.width}
        style={{
          top: top + this.props.position.row,
          left: left + this.props.position.column,
        }}
        ref={this.props.forwardedRef}
      ></Cursor>
    );
  }
}

export default CursorLayer;
