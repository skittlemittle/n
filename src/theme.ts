import { COLORS_L } from "./themeConstants";

export const AppTheme = {
  colors: COLORS_L,
};
