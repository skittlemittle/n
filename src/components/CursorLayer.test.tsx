import renderer from "react-test-renderer";
import { ThemeProvider } from "styled-components";
import { AppTheme } from "../theme";
import CursorLayer from "./CursorLayer";

it("cursor block", () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={AppTheme}>
        <div id="gaming">
          <CursorLayer
            position={{ row: 300, column: 300 }}
            size={{ height: 19, width: 6 }}
            parent={"gaming"}
            block={true}
          />
        </div>
      </ThemeProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("cursor line", () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={AppTheme}>
        <div id="gaming">
          <CursorLayer
            position={{ row: 300, column: 300 }}
            size={{ height: 19, width: 6 }}
            parent={"gaming"}
            block={false}
          />
        </div>
      </ThemeProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
