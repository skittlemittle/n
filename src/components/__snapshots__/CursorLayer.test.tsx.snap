// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cursor block 1`] = `
<div
  id="gaming"
>
  <div
    className="sc-bdvvtL laJtQv"
    hidden={true}
    style={
      Object {
        "left": 300,
        "top": 300,
      }
    }
  />
</div>
`;

exports[`cursor line 1`] = `
<div
  id="gaming"
>
  <div
    className="sc-bdvvtL knxTwh"
    hidden={true}
    style={
      Object {
        "left": 300,
        "top": 300,
      }
    }
  />
</div>
`;
